File: src/main/java/com/freya02/bot/commands/text/Exit.java
package com.freya02.bot.commands.text;

import com.freya02.botcommands.api.Logging;
import com.freya02.botcommands.api.annotations.RequireOwner;
import com.freya02.botcommands.api.prefixed.BaseCommandEvent;
import com.freya02.botcommands.api.prefixed.TextCommand;
import com.freya02.botcommands.api.prefixed.annotations.JDATextCommand;
import net.dv8tion.jda.api.JDA;
import org.slf4j.Logger;

public class Exit extends TextCommand {
	private static final Logger LOGGER = Logging.getLogger();

	@RequireOwner
	@JDATextCommand(name = "exit")
	public void say(BaseCommandEvent event) throws InterruptedException {
		LOGGER.warn("Shutdown initiated by {} ({})", event.getAuthor().getAsTag(), event.getAuthor().getId());

		event.reactSuccess().mapToResult().complete();

		event.getJDA().shutdown();

		while (event.getJDA().getStatus() != JDA.Status.SHUTDOWN) {
			Thread.sleep(15);

			Thread.onSpinWait();
		}

		System.exit(0);
	}
}

File: src/main/kotlin/com/freya02/bot/commands/text/Exit.kt
package com.freya02.bot.commands.text

import dev.minn.jda.ktx.coroutines.await
import io.github.freya022.botcommands.api.commands.annotations.Command
import io.github.freya022.botcommands.api.commands.text.BaseCommandEvent
import io.github.freya022.botcommands.api.commands.text.TextCommand
import io.github.freya022.botcommands.api.commands.text.annotations.JDATextCommandVariation
import io.github.freya022.botcommands.api.commands.text.annotations.RequireOwner
import io.github.freya022.botcommands.api.core.utils.awaitShutdown
import io.github.oshai.kotlinlogging.KotlinLogging
import kotlin.system.exitProcess
import kotlin.time.Duration.Companion.seconds

@Command
class Exit : TextCommand() {
    private val logger = KotlinLogging.logger { }

    @RequireOwner
    @JDATextCommandVariation(path = ["exit"])
    suspend fun exit(event: BaseCommandEvent) {
        try {
            logger.warn { "Shutdown initiated by ${event.author.name} (${event.author.id})" }
            event.reactSuccess().mapToResult().await()

            event.jda.shutdown()
            event.jda.awaitShutdown(15.seconds)
        } catch (e: Exception) {
            exitProcess(2)
        } finally {
            exitProcess(0)
        }
    }
}